### PR TITLE
Add setInputFiles to ElementHandle

### PR DESCRIPTION
**Pull request by bandorko**
## What?

Add `setInputFiles` support.

## Test
Here is a test script you can test this functionality with

**index.html:** (served by e.g. ```python3 -m http.server 3080```)
```
<html>

<head>
    <script src="https://unpkg.com/dropzone@5/dist/min/dropzone.min.js"></script>
</head>

<body>
    <!-- Simple file upload form -->
    <form method="POST" action="/upload" enctype="multipart/form-data">
        <input type="file" name="upl" id="simple-upload" multiple />
        <input type="submit" value="Send" />
    </form>

    <!-- Dropzone.js -->
    <div id="dropzone"/>
    <script>
        let myDropzone = new Dropzone("div#dropzone", { url: "/upload", autoProcessQueue:false });
    </script>
</body>

</html>
```

**test.js:**
```
import { browser } from 'k6/experimental/browser';
import { check } from 'k6';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      vus: 1,
      iterations: 1,
      options: {
        browser: {
          type: 'chromium',
        },
      },
    },
  }
};

function getFileCount(page, id) {
  return page.evaluate((id) => {
    return document.getElementById(id).files.length;
  }, id);
}

function getFileProp(page, id, fileNo, prop) {
  return page.evaluate((id, fileNo, prop) => {
    return document.getElementById(id).files[fileNo][prop];
  }, id, fileNo, prop);
}

function getDropzoneFileCount(page) {
  return page.evaluate(() => {
    return myDropzone.files.length;
  });
}

function getDropzoneFileProp(page, fileNo, prop) {
  return page.evaluate((fileNo, prop) => {
    return myDropzone.files[fileNo][prop];
  }, fileNo, prop);
}

export default async function () {
  const page = browser.newPage();
  try {
    await page.goto('http://localhost:3080/');

    // test simple upload form
    page.setInputFiles('input[id="simple-upload"]', { name: 'test.json', mimetype: 'text/json', buffer: 'MDEyMzQ1Njc4OQ==' });
    check(getFileCount(page, 'simple-upload'), { 'file count is 1': (count) => count === 1 });
    check(getFileProp(page, 'simple-upload', 0, 'name'), { '1st file name is test.json': (prop) => prop === 'test.json' });
    check(getFileProp(page, 'simple-upload', 0, 'type'), { '1st file mimetype is text/json': (prop) => prop === 'text/json' });
    check(getFileProp(page, 'simple-upload', 0, 'size'), { '1st file size is 10': (prop) => prop === 10 });

    page.setInputFiles('input[id="simple-upload"]', [{ name: 'test.json', mimetype: 'text/json', buffer: 'MDEyMzQ1Njc4OQ==' }, { name: 'hello.txt', mimetype: 'text/plain', buffer: 'SGVsbG8gSzYgV29ybGQK' }]);
    check(getFileCount(page, 'simple-upload'), { 'file count is 2': (count) => count === 2 });
    check(getFileProp(page, 'simple-upload', 0, 'name'), { '1st file name is test.json': (prop) => prop === 'test.json' });
    check(getFileProp(page, 'simple-upload', 0, 'type'), { '1st file mimetype is text/json': (prop) => prop === 'text/json' });
    check(getFileProp(page, 'simple-upload', 0, 'size'), { '1st file size is 10': (prop) => prop === 10 });
    check(getFileProp(page, 'simple-upload', 1, 'name'), { '2nd file name is hello.txt': (prop) => prop === 'hello.txt' });
    check(getFileProp(page, 'simple-upload', 1, 'type'), { '2nd file mimetype is text/plain': (prop) => prop === 'text/plain' });
    check(getFileProp(page, 'simple-upload', 1, 'size'), { '2nd file size is 15': (prop) => prop === 15 });

    page.setInputFiles('input[id="simple-upload"]', null);
    check(getFileCount(page, 'simple-upload'), { 'file count is 0': (count) => count === 0 });

    page.setInputFiles('input[id="simple-upload"]', 'test.js');
    check(getFileCount(page, 'simple-upload'), { 'file count is 1': (count) => count === 1 });
    check(getFileProp(page, 'simple-upload', 0, 'name'), { '1st file name is test.js': (prop) => prop === 'test.js' });
    check(getFileProp(page, 'simple-upload', 0, 'type'), { '1st file mimetype is text/javascript; charset=utf-8': (prop) => prop === 'text/javascript; charset=utf-8' });
    check(getFileProp(page, 'simple-upload', 0, 'size'), { '1st file size is larger than 1000': (prop) => prop > 1000 });

    page.setInputFiles('input[id="simple-upload"]', ['test.js', 'test.js']);
    check(getFileCount(page, 'simple-upload'), { 'file count is 2': (count) => count === 2 });

    // test dropzone.js form
    page.setInputFiles('input[class="dz-hidden-input"]', { name: 'test.json', mimetype: 'text/json', buffer: 'MDEyMzQ1Njc4OQ==' });
    check(getDropzoneFileCount(page), { 'dropzone file count is 1': (count) => count === 1 });
    check(getDropzoneFileProp(page, 0, 'name'), { 'dropzone 1st file name is test.json': (prop) => prop === 'test.json' });
    check(getDropzoneFileProp(page, 0, 'type'), { 'dropzone 1st file mimetype is text/json': (prop) => prop === 'text/json' });
    check(getDropzoneFileProp(page, 0, 'size'), { 'dropzone 1st file size is 10': (prop) => prop === 10 });
  } finally {
    page.close();
  }
}

```

## Why?

Users can select input files for upload.

## Checklist

- [ ] I have performed a self-review of my code
- [ ] I have added tests for my changes
- [ ] I have commented on my code, particularly in hard-to-understand areas

## Related PR(s)/Issue(s)

Closes: #384

### CODE REVIEW

**Comment by bandorko** [comment]
I know that this is far from ready, just please give me feedback, if I am on the right track?

It works like this:
```
import { browser } from 'k6/experimental/browser';
import { sleep } from 'k6';
import encoding from 'k6/encoding';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      vus: 1,
      iterations: 1,
      options: {
        browser: {
          type: 'chromium',
        },
      },
    },
  }
}

export default async function () {
  const page = browser.newPage();
  try {
    await page.goto('http://localhost:9080');
    const input = page.$('input[name="upload"]');
    input.setInputFiles({ payload: [{ name: 'filename.txt', buffer: encoding.b64encode('this is test'), mime: 'text/plain' },{ name: 'filename2.txt', buffer: encoding.b64encode('this is second test'), mime: 'text/plain' }] });
    sleep(20);
  } finally {
    page.close();
  }
}
```

with this html:
```
<html>
    <input type="file" name="upload"/>
</html>
```

File: common/element_handle_options.go
Comment (inancgumus) [inline common/element_handle_options.go:198]: You don't need to initialize this (`[]File{}`) here. It's automatic.

File: common/element_handle_options.go
Comment (inancgumus) [inline common/element_handle_options.go:208]: Can you early-return here?

File: common/element_handle_options.go
Comment (inancgumus) [inline common/element_handle_options.go:211]: Instead of nesting, can you also use early-continue here? Like:

`if k != "payload" { continue }`

**Review by inancgumus — COMMENTED** [review]
Thanks, @bandorko. It looks LGTM so far 👏 Added some suggestions. We also need more tests to ensure that this works. Please also use our commit message conventions and add descriptions to your commits. You might also want to consider [this issue](https://github.com/grafana/xk6-browser/issues/355) next :)

File: common/element_handle_options.go
Comment (bandorko) [reply common/element_handle_options.go]: done

File: common/element_handle_options.go
Comment (bandorko) [reply common/element_handle_options.go]: done

File: common/element_handle_options.go
Comment (bandorko) [reply common/element_handle_options.go]: done

**Comment by bandorko** [comment]
@inancgumus 
I implemented SetInputFiles in Frame and Page also and I created tests for the functionality.

File: common/element_handle.go
Comment (inancgumus) [inline common/element_handle.go:1243]:
```suggestion
// SetInputFiles sets the given files into the input file element.
```

File: common/element_handle.go
Comment (inancgumus) [inline common/element_handle.go:1267]:
```suggestion
		if strings.TrimSpace(file.Path) != "" {
```

File: common/element_handle.go
Comment (inancgumus) [inline common/element_handle.go:1270]:
```suggestion
				return fmt.Errorf("reading file: %w", err)
```

File: common/element_handle_options.go
Comment (inancgumus) [inline common/element_handle_options.go:218]:
```suggestion
			return fmt.Errorf("parsing file descriptor: %w", err)
```

File: common/element_handle_options.go
Comment (inancgumus) [inline common/element_handle_options.go:224]:
```suggestion
		return fmt.Errorf("invalid parameter type %T", fileType.Kind())
```

File: common/frame.go
Comment (inancgumus) [inline common/frame.go:1512]: It would be great if we could return an error from this function instead of panicking.

File: tests/page_test.go
Comment (inancgumus) [inline tests/page_test.go:1112]: Could you rename all test names in this format:

```suggestion
			name: "set_one_file_with_object",
```

**Review by inancgumus — COMMENTED** [review]
Mostly LGTM 👏 Thanks for the changes. I've added some suggestions.

File: common/frame.go
Comment (ankur22) [inline common/frame.go:1506]: Can you change the comment for this method?

**Review by ankur22 — CHANGES_REQUESTED** [review]
Thank you for the contribution! 🎉 This should be quite useful for everyone when they wish to test uploading of files from their websites 🙇 

I've left some comments for you to ponder over and work on if you think they are valid.

File: common/element_handle.go
Comment (bandorko) [reply common/element_handle.go]: Do you think something like this? https://github.com/grafana/xk6-browser/pull/1097/commits/15f627d349133f20395f0a93059690b3102b4406

File: common/js/injected_script.js
Comment (bandorko) [reply common/js/injected_script.js:437]: I just followed the pattern in this file. Every occurance is "error:notelement"

File: common/element_handle.go
Comment (ankur22) [reply common/element_handle.go]: Yep, thanks!

File: common/js/injected_script.js
Comment (bandorko) [reply common/js/injected_script.js:436]: It is possible, I also followed the pattern in the file, but I will rewrite it

File: common/js/injected_script.js
Comment (ankur22) [reply common/js/injected_script.js:437]: @inancgumus wdyt is the best way forward with the errors in the injected script?

Comment (bandorko) [reply common/js/injected_script.js:437]: Maybe I should rewrite the other two error to not contain spaces ('error:not input element' and 'error:Not an input[type=file] element')

File: common/element_handle_options.go
Comment (bandorko) [reply common/element_handle_options.go]: done

File: common/element_handle.go
Comment (bandorko) [reply common/element_handle.go]: done

File: common/element_handle.go
Comment (bandorko) [reply common/element_handle.go]: done

File: common/element_handle.go
Comment (bandorko) [reply common/element_handle.go]: done

File: common/element_handle_options.go
Comment (ankur22) [inline common/element_handle_options.go:224]: Wondering if we need to return an error when `ok` is `false`?

I know i'm being pedantic here since the `if` statement is unnecessary here and only here due to the linter.

Happy for you to ignore this comment.

**Review by ankur22 — DISMISSED** [review]
Awesome work 👏 thanks for the contribution!

I'm wondering if you could add the details of a test script that you've used to test this change in the PR description for posterity reasons?

I've also left a nit-pick which i'm very happy for you to ignore.

**Review by ka3de — DISMISSED** [review]
Great job @bandorko ! Thank you for your contribution.
LGTM after the existing conflicts are resolved.

File: common/js/injected_script.js
Comment (ankur22) [reply common/js/injected_script.js:437]: Don't make this change i've suggested? I've just found that we're relying on `error:notelement` [here](https://github.com/grafana/xk6-browser/blob/main/common/element_handle.go#L1547). Thanks!

**Review by inancgumus — DISMISSED** [review]
Great work and initiative. Thanks for your contribution 🙇

File: common/js/injected_script.js
Comment (bandorko) [reply common/js/injected_script.js:437]: Thanks. Maybe I should then use error:notinput instead of error:notinputelement in line 439, and add error:notinputfileelement to the map in element_handle.go. @ankur22 what do you think?

Comment (ankur22) [reply common/js/injected_script.js:437]: Yep, i like that idea 👍

**Comment by bandorko** [comment]
@ankur22 , @ka3de , @inancgumus : Thank you!
I still want to fix the issues @ankur22 found, and there are still lint issues I don't know how to handle:
```
common/element_handle_options.go:214:2  exhaustive  missing cases in switch of type reflect.Kind: reflect.Invalid, reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr, reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.Array, reflect.Chan, reflect.Func, reflect.Interface, reflect.Pointer|reflect.Ptr, reflect.Slice, reflect.Struct, reflect.UnsafePointer
common/element_handle_options.go:240:2  exhaustive  missing cases in switch of type reflect.Kind: reflect.Invalid, reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr, reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.Array, reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Pointer|reflect.Ptr, reflect.String, reflect.Struct, reflect.UnsafePointer
```

**Comment by inancgumus** [comment]
Hi @bandorko,

> there are still lint issues I don't know how to handle

What difficulties are you having with the linter issues? We can help.

**Comment by ankur22** [comment]
@bandorko I think it might be ok to add `//nolint: exhaustive` for this particular case. WDYT @inancgumus?

**Comment by bandorko** [comment]
> @bandorko I think it might be ok to add `//nolint: exhaustive` for this particular case. WDYT @inancgumus?

There is an other option you may want to consider: setting default-signifies-exhaustive: true for exhaustive linter

File: common/element_handle_options.go
Comment (bandorko) [reply common/element_handle_options.go:224]: I don't see any chance that the type assertion would fail if the file.exportType() is reflect.String, so I think it would be an unreachable, untestable code.

**Comment by inancgumus** [comment]
> There is an other option you may want to consider: setting default-signifies-exhaustive: true for exhaustive linter

Thanks, @bandorko, for the suggestion. But I'm on the side of keeping the exhaustive linter as it allows us to catch bugs early. WDYT @ankur22?

**Comment by ankur22** [comment]
> Thanks, @bandorko, for the suggestion. But I'm on the side of keeping the exhaustive linter turned on (w/o a default turn off-er) as it allows us to catch bugs early. WDYT @ankur22?

Agreed 👍 Let's keep it the same for now and review each on a case-by-case basis.

**Comment by bandorko** [comment]
@inancgumus , @ankur22 I fixed the exhaustive lint issue as you suggested. What else can I do for this PR?

**Comment by inancgumus** [comment]
Hi @bandorko, we can merge this PR after fixing the conflicts with the `main` branch. Please let us know when you're done. Thanks!

**Comment by bandorko** [comment]
@inancgumus 
I added a test script to the description as @ankur22 asked a month ago, so I think I am done with this PR.

**Comment by inancgumus** [comment]
Hi @bandorko,

> I added a test script to the description as @ankur22 asked a month ago, so I think I am done with this PR.

That's great. Would you like to resolve the conflicts? Or we can take it over from here? WDYT?

> This branch has conflicts that must be resolved
> Use the [web editor](https://github.com/grafana/xk6-browser/pull/1097/conflicts) or the  to resolve conflicts.
> Conflicting files
>tests/page_test.go

**Comment by bandorko** [comment]
As I see, it is not just a simple merge conflict, because there were refactors recently. So I will try to resolve the issues.

**Comment by inancgumus** [comment]
> As I see, it is not just a simple merge conflict, because there were refactors recently. So I will try to resolve the issues.

Hey @bandorko. These refactorings shouldn't affect the functionality offered by this PR. Please let us know how it goes. Good luck 🤞

**Comment by ankur22** [comment]
Hey @bandorko, how are you getting on with this change? Can we help with the rebase? We're planning on to perform a release later this week and have it part of the latest version of k6.

**Comment by bandorko** [comment]
I resolved the conflict and merged the main branch to this feature branch. I built it, and it still works on my computer with the test scenario in the description, but **make lint** gives me some error in files I didn't even touch.

File: common/js/injected_script.js
Comment (bandorko) [reply common/js/injected_script.js:437]: I have done this also

**Review by inancgumus — APPROVED** [review]
@bandorko Thank you for your efforts! 🎉🙇

**Review by ankur22 — DISMISSED** [review]
LGTM 🚀

**Comment by inancgumus** [comment]
Hi @bandorko, I'm going to arrange the commits in this PR and rebase it to `main`, and then merge the PR.

**Comment by inancgumus** [comment]
I can't merge this PR due to this message:

"This branch cannot be rebased due to conflicts"

@bandorko, can I work on your personal fork (`bandorko:feature/384-setInputFiles`) and push there first to fix the issues on your behalf?

**Comment by bandorko** [comment]
@inancgumus Yes, of course. Should I give you permission for my fork?

**Comment by inancgumus** [comment]
> @inancgumus Yes, of course. Should I give you permission for my fork?

@bandorko Yes, please, thanks!

**Comment by inancgumus** [comment]
@bandorko I couldn't resolve the conflicts. There are conflicts to resolve when we do:

```
# in your fork
git checkout main
git pull upstream main
gc feature/384-setInputFiles
git rebase main
```

**Review by inancgumus — APPROVED** [review]

**Review by ankur22 — APPROVED** [review]

**Comment by erashdan** [comment]
worked for me thanks ! 👍🏻

**Pull request merged**
